### DISCUSSION TITLE
Doubts regarding this build

### DISCUSSION THREAD

**Discussion by JRM-PT** [mwood77/winderoo#43] [General]
Hi,

I was looking for an upgrade for my AliExpress cheap watch winder and found your project. I liked it a lot, especially the browser UI and the way it enables a quick setup.
I don't need the case and watch cradle (that I can reuse from the AliExpress one) but the electronics I want to replace as the controls are terrible and the motor is noisy (probably the plastic gears).
I might change the motor by a stepper motor (need to change part of the code but not that difficult). Do you know a motor driver that works silently?
Another question is regarding the addition of a start/stop button: placed between the motor driver and the motor; this will stop the motor but not the ESP32, that will still count the cycle running time. It's useful to remove the watch from the winder but not for stop/starting the cycle. It would be much more useful if one could start/stop/restart the cycle with a push button (I think I saw something similar in another project).
Finally, I wish to add a LCD/OLED display that reveals the Status, Turns since start and Time to end of cycle, all variables that are already sent to the web UI. Do you think it's possible without rewriting too much code?
Thanks

**Comment by mwood77** [comment]
Hi @JRM-PT 

Great questions - I'll do my best to answer them succintly.

> I might change the motor by a stepper motor (need to change part of the code but not that difficult). Do you know a motor driver that works silently?

I haven't explored this, so I don't know. However there's the possibility that we could add an optical rotary encoder instead to track the position of the winding unit (or the motor's position if you know the gear ratio). The assumption being that's the reason why you'd prefer a stepper motor of a DC motor.


> Another question is regarding the addition of a start/stop button: placed between the motor driver and the motor; this will stop the motor but not the ESP32, that will still count the cycle running time. It's useful to remove the watch from the winder but not for stop/starting the cycle. It would be much more useful if one could start/stop/restart the cycle with a push button (I think I saw something similar in another project).

Right, in the [OSWW wiring diagram](https://github.com/mwood77/osww/blob/main/instructions/wiring-diagram.md) there's a physical switch which cuts power to the motor. In [Winderoo, I've removed that switch](https://github.com/mwood77/winderoo/blob/main/docs/wiring-diagram.md) as it's now software controlled. Depending on what you've got on-hand, you can also do this without using Winderoo's UI. For example, you could have HomeAssistant toggle the winder's state on when a (zigbee?) button is pressed. All you'd have to do is send a call Winderoo's ESP32 over your network.

For example:

1. Start by getting the state of the winder
```
GET http://winderoo.local/api/status so you can get the current state of the winder
// Response body from the winder
      {
        "status": "Stopped",               <------
        "rotationsPerDay": "300",
        "direction": "BOTH",
        "hour": "12",
        "minutes": "50",
        "durationInSecondsToCompleteOneRevolution": 8,
        "startTimeEpoch": 0,
        "currentTimeEpoch": 1680555863,
        "estimatedRoutineFinishEpoch": 0,
        "winderEnabled": "1",
        "timerEnabled": "1",
        "db": -67
      }
```

2. Then you could send an update request to start winding it:
```
POST http://winderoo.local/api/update 
// Request body
      {
        "tpd": "300",
        "hour": "12",
        "minutes": "50",
        "timerEnabled": "1",
        "action": "START",               <------
        "rotationDirection": "BOTH"
      }
```

If you do this, it's important to input the correct data for `tpd`, `hour`, `minutes`, `timerEnabled`, and `rotationDirection`. All of these you can get from the first step. To stop it, send the same request to `POST  http://winderoo.local/api/update` but make the action field, `"action": "STOP"`

> Finally, I wish to add a LCD/OLED display that reveals the Status, Turns since start and Time to end of cycle, all variables that are already sent to the web UI. Do you think it's possible without rewriting too much code?

This is a bit of a bigger ask, as I'd have to add in LCD drivers to the backend and then handle drawing (parts of) the screen. It isn't super difficult, but I'm curious how it would impact the available memory of the board. We'd also need to introduce some kind of build flag to make this part optional.

**Reply by JRM-PT** [reply]
Thanks for your answers.
The reason to change the motor is because it seats on my bedside table and the cheap plastic gears are too noisy (winning noise is quite annoying and I tend to hear all this kind of defects). I saw a guy using a TMC2100 based driver claiming it was quite silent but with ATMEGA328 board, not ESP32.
If I disconnect the power  supply (phone charger), after I reconnect it and the ESP to reboot, it will start running the code by itself?

**Reply by mwood77** [reply]
Winderoo saves its state to the ESP32's NVRAM, so if power is lost, it'll resume to its last known state.

If it was winding, it'll begin winding again (although back at 0% - I don't save the winding % completed). This also means it'll reconnect to the last known wifi network.

**Comment by JRM-PT** [comment]
Looks like TMC2209 is a good silent driver for stepper motors and ESP32.

**Reply by JRM-PT** [reply]
Just bought the ESP32 board (that has a small OLED attached) plus a TM2209 driver and a simple stepper motor. I'll start by the standard Winderoo with the old motor. Then, I'll progressively try to increase the complexity by adding the other features like the better controlled stepper motor and the mirror of the settings/status to the OLED.
I might need help 😂.
Do you already have Portuguese as a possible language? If not, I can translate what you need

**Reply by mwood77** [reply]
Do you have a link to the ESP32 board that you bought?

You could also attach a separate LED + resistor to another pin on the ESP32; changing the code to output the LED states on a different pin is very easy and the code can already do this. All you have to do is change the GPIO pin here: https://github.com/mwood77/winderoo/blob/3e510b48939804cb87acaadf1c19954f7a6b4280/src/platformio/osww-server/src/main.cpp#L34

---

RE: Portuguese

You mean as an available language in the UI? I don't, but I'd love to add it!

These are the translatable items - if you could translate the blue text, I'll add it in:

```JSON
{
    "HEADER": {
        "RESET": "Reset",
        "ON": "on",
        "OFF": "off",
        "DIALOG": {
            "TITLE": "Reset takes 20 seconds",
            "SUBTITLE": "This screen will not close during the reset process",
            "CANCEL": "cancel",
            "POINT_1": "Reset is complete when the second LED stops fast blinking, then turns solid.",
            "POINT_2": "Connect your mobile device to the new WiFi network called",
            "POINT_3": "Add Winderoo to an existing WiFi network. If the LED starts to slow blink, Winderoo has successfully joined that network.",
            "POINT_4": "Wait for your device to reconnect to your existing WiFi network.",
            "NOW_AVAILABLE": "Winderoo is now available at:",
            "CONFIRM_RESET": "Confirm Reset"
        }
    },
    "SETTINGS": {
        "STATUS": "Status",
        "STOPPED": "Stopped",
        "WINDING": "Winding",
        "ESTIMATED_DURATION": "Estimated Cycle Duration",
        "DIRECTION": "Direction",
        "ROTATIONS_PER_DAY": "Rotations Per Day",
        "FIND_WINDING_PARAMETERS": "Find your watch's winding parameters",
        "CYCLE_START": "Cycle Start Time",
        "ENABLED": "Enabled",
        "DISABLED": "Disabled",
        "HOURS": "Hour",
        "MINUTES": "Minutes",
        "SAVE": "Save Settings",
        "SAVE_IN_PROGRESS": "Upload in progress",
        "CLOCKWISE": "Clockwise",
        "COUNTER_CLOCKWISE": "Counter Clockwise",
        "BOTH": "Both",
        "PLEASE_WAIT": "Getting cycle progress, please wait",
        "PROGRESS": "Cycle Progress"
    }
}
```

**Reply by JRM-PT** [reply]
It was this one:
https://a.aliexpress.com/_Ew9TAXB
I was offering to translate things to Portuguese. Haven't you asked for contributions in order to have more languages available?

**Reply by mwood77** [reply]
I have! But I didn't initially understand what you were asking...but then got it! 

If you look at my previous message, I've posted the translatable items 😃. I'd love to add Portuguese in!

-------

> It was this one:
https://a.aliexpress.com/_Ew9TAXB

Interesting - I haven't used that board, but I don't see why it wouldn't work. If I've got some time in the coming weeks, I might be able to experiment with driving an OLED. 

Come to think of it, I might have a spare OLED screen around the house...

**Reply by mwood77** [reply]
@JRM-PT I've added early OLED support to Winderoo in #47 , but it's still a few hours work away from release. If you have your board, you can flash that working branch onto it.

**Reply by mwood77** [reply]
OLED support has been merged-in; pull the `main` branch, and you'll have it.

**Comment by mwood77** [comment]
@JRM-PT This pull request #44  adds Português to Winderoo. If you could reply there with the translations, that would be great!
